velox/exec/tests/TestValueTest.cpp

```diff
@@ -0,0 +1,230 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <gtest/gtest.h>
+
+#include "velox/exec/tests/utils/TestValue.h"
+
+namespace facebook::velox::exec::test {
+class TestObject {
+ public:
+  TestObject() = default;
+  ~TestObject() = default;
+
+  void set(int value) {
+    TestValue::adjust("facebook::velox::exec::test::TestObject::set", &value);
+    internalSet(value);
+  }
+
+ private:
+  void internalSet(int value) {
+    if (TestValue::enabled()) {
+      std::pair<int, int> testValue(value, count_);
+      TestValue::adjust(
+          "facebook::velox::exec::test::TestObject::internalSet", &testValue);
+    }
+    ++count_;
+    value_ = value;
+  }
+
+  int count_ = 0;
+  int value_;
+};
+
+TEST(TestValueTest, basic) {
+  for (const bool enabled : {false, true}) {
+    SCOPED_TRACE(fmt::format("testvalue enabled: {}", enabled));
+    if (enabled) {
+      TestValue::enable();
+    } else {
+      TestValue::disable();
+    }
+    EXPECT_EQ(enabled, TestValue::enabled());
+    {
+      int setCount = 0;
+      int setValue = 0;
+      int value = 200;
+      int changeValue = 100;
+      TestValue::set(
+          "facebook::velox::exec::test::TestObject::set",
+          [&](void* testData) {
+            int* valueToChange = static_cast<int*>(testData);
```

Comment:
It is cumbersome to work with void*. Would it make sense for TestObject to provide a helper API that takes a fully-typed callback?, e.g. void(int value)

Reply:
The testData type could be different for different test cases. And we can't put the registered callback in one map. So do you mean TestObject provide a helper API and we do the conversion in TestValue::set here? Thanks!

Reply:
I make the set method a template one. Thanks for the suggestion!